File: Kalorie/Kalorie/Core/UseCases/SubmitFoodItemReportUseCase.swift
//
//  SubmitFoodItemReportUseCase.swift
//  Kalorie
//
//  Created by Josef Antoni on 17.09.2026.
//

import Foundation

protocol SubmitFoodItemReportUseCaseProtocol {
    func callAsFunction(barcode: String, reason: String) async throws
}

struct SubmitFoodItemReportUseCase: SubmitFoodItemReportUseCaseProtocol {

    // MARK: - Properties

    private let dataProvider: any FirestoreDataProviderProtocol
    private let authProvider: any AuthProviderProtocol

    // MARK: - Init

    init(dataProvider: any FirestoreDataProviderProtocol, authProvider: any AuthProviderProtocol) {
        self.dataProvider = dataProvider
        self.authProvider = authProvider
    }

    // MARK: - Functions

    func callAsFunction(barcode: String, reason: String) async throws {
        guard let userId = authProvider.userId else { throw AuthError.notAuthenticated }
        let trimmedReason = reason.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmedReason.isEmpty else { throw FoodItemReportError.reasonRequired }
        // Rules' string.size() counts UTF-16 code units — an emoji is one Character but two or more units.
        guard trimmedReason.utf16.count <= Constants.Firestore.reportReasonMaxLength else {
            throw FoodItemReportError.reasonTooLong
        }
        let dto = FoodItemReportDTO(barcode: barcode, reportedBy: userId, reason: trimmedReason, reportedAt: .now)
        try await dataProvider.setAsync(
            dto,
            id: FoodItemReportDomain.id(barcode: barcode, userId: userId),
            in: Constants.Firestore.foodItemReports
        )
    }
}

#if DEBUG
struct SubmitFoodItemReportUseCaseFake: SubmitFoodItemReportUseCaseProtocol {

    // MARK: - Properties

    var errorToThrow: Error?

    // MARK: - Functions

    func callAsFunction(barcode: String, reason: String) async throws {
        if let errorToThrow { throw errorToThrow }
    }
}
#endif

File: Android/app/src/main/java/antoni/kalorie/core/usecases/SubmitFoodItemReportUseCase.kt
package antoni.kalorie.core.usecases

import antoni.kalorie.core.auth.AuthError
import antoni.kalorie.core.auth.AuthProviderProtocol
import antoni.kalorie.core.models.FoodItemReportDomain
import antoni.kalorie.core.models.FoodItemReportError
import antoni.kalorie.core.networking.FirestoreDataProviderProtocol
import antoni.kalorie.core.networking.FoodItemReportDTO
import antoni.kalorie.core.networking.setAsync
import antoni.kalorie.core.utils.Constants
import java.time.Instant

interface SubmitFoodItemReportUseCaseProtocol {
    suspend operator fun invoke(barcode: String, reason: String)
}

class SubmitFoodItemReportUseCase(
    private val dataProvider: FirestoreDataProviderProtocol,
    private val authProvider: AuthProviderProtocol,
) : SubmitFoodItemReportUseCaseProtocol {

    // MARK: - Functions

    override suspend fun invoke(barcode: String, reason: String) {
        val userId = authProvider.userId ?: throw AuthError.NotAuthenticated
        val trimmedReason = reason.trim()
        if (trimmedReason.isEmpty()) throw FoodItemReportError.ReasonRequired
        if (trimmedReason.length > Constants.Firestore.REPORT_REASON_MAX_LENGTH) throw FoodItemReportError.ReasonTooLong
        val dto = FoodItemReportDTO(barcode = barcode, reportedBy = userId, reason = trimmedReason, reportedAt = Instant.now())
        dataProvider.setAsync(
            dto,
            id = FoodItemReportDomain.id(barcode = barcode, userId = userId),
            inCollection = Constants.Firestore.FOOD_ITEM_REPORTS,
        )
    }
}
